ledger/complete/compactor.go

```diff
@@ -0,0 +1,393 @@
+	// - tries snapshot fails.
+	// NOTE: next checkpoint number must >= active segment num.
+	// We can't reuse mtrie state to checkpoint tries in older segments.
+	nextCheckpointNum := lastCheckpointNum + int(c.checkpointDistance)
+	if activeSegmentNum > nextCheckpointNum {
+		nextCheckpointNum = activeSegmentNum
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+
+Loop:
+	for {
+		select {
+
+		case doneCh := <-c.stopCh:
+			defer close(doneCh)
+			cancel()
+			break Loop
+
+		case checkpointResult := <-checkpointResultCh:
+			if checkpointResult.err != nil {
+				c.logger.Error().Err(checkpointResult.err).Msgf(
+					"compactor failed to checkpoint %d", checkpointResult.num,
+				)
+
+				// Retry checkpointing after active segment is finalized.
+				nextCheckpointNum = activeSegmentNum
+			}
+
+		case update, ok := <-c.trieUpdateCh:
+			if !ok {
+				// trieUpdateCh channel is closed.
+				// Wait for stop signal from c.stopCh
+				continue
+			}
+
+			var checkpointNum int
+			var checkpointTries []*trie.MTrie
+			activeSegmentNum, checkpointNum, checkpointTries =
+				c.processTrieUpdate(update, activeSegmentNum, nextCheckpointNum)
+
+			if checkpointTries == nil {
+				// Don't checkpoint yet because
+				// - not enough segments for checkpointing (nextCheckpointNum >= activeSegmentNum), or
+				// - failed to get ledger state snapshop (nextCheckpointNum < activeSegmentNum)
+				if nextCheckpointNum < activeSegmentNum {
+					nextCheckpointNum = activeSegmentNum
+				}
+				continue
+			}
+
+			// Try to checkpoint
+			if checkpointSem.TryAcquire(1) {
+
+				// Compute next checkpoint number
+				nextCheckpointNum = checkpointNum + int(c.checkpointDistance)
+
+				go func() {
+					defer checkpointSem.Release(1)
+					err := c.checkpoint(ctx, checkpointTries, checkpointNum)
+					checkpointResultCh <- checkpointResult{checkpointNum, err}
+				}()
+			} else {
+				// Failed to get semaphore because checkpointing is running.
+				// Try again when active segment is finalized.
+				c.logger.Info().Msgf("compactor delayed checkpoint %d because prior checkpointing is ongoing", nextCheckpointNum)
+				nextCheckpointNum = activeSegmentNum
+			}
+		}
+	}
+
+	// Drain and process remaining trie updates in channel.
+	for update := range c.trieUpdateCh {
+		_, _, err := c.wal.RecordUpdate(update.Update)
+		select {
+		case update.ResultCh <- err:
+		default:
+		}
+	}
+
+	// Don't wait for checkpointing to finish because it might take too long.
+}
+
+// processUpdateResult sends WAL update result using ResultCh channel
+// and waits for signal from DoneCh channel.
+// This ensures that WAL update and ledger state update are in sync.
+func processUpdateResult(update *WALTrieUpdate, updateResult error) {
+	// Send result of WAL update
+	update.ResultCh <- updateResult
+
+	// Wait for trie update to complete
+	<-update.DoneCh
+}
+
+func (c *Compactor) checkpoint(ctx context.Context, tries []*trie.MTrie, checkpointNum int) error {
```

Comment:
checkpoint doesn't happen often. It worth to add logs about when checkpointing is started, and when it's finished, as well as when it's aborted or failed.

Reply:
The checkpoint logging is already done in [`createCheckpoint()`](https://github.com/onflow/flow-go/blob/b8797cc2e4564ffceae8cde1bb577cf2fd0863ba/ledger/complete/compactor.go#L274-L301).